### PR TITLE
FP8 Enablement for TE layers + HSTU attn

### PR DESCRIPTION
**Pull request by esoba**
Added MixedPrecisionArgs to allow users to configure FP8 usage in TE linear layers and HSTU attention for Native HSTU layer. Features include:
- New MixedPrecisionArgs in gin config
- Support for TE fp8 autocast in pipeline
- Ability to separate TE FP8 from HSTU FP8 attn 
- Truncation/padding logic to enable TE FP8 divisible by 16 requirement 

Minimal working example PYTHONPATH=${PYTHONPATH}:$(realpath ../) torchrun --nproc_per_node 2 --master_addr localhost --master_port 6000  pretrain_gr_ranking.py --gin-config-file movielens_ranking_fp8.gin

Setup currently has a bug when both TE linear layer and HSTU attn are fp8 enabled, seeing NaN loss at iteration 64. I have a debugging branch [here](https://github.com/esoba/recsys-examples/tree/debugging-with-repro) that tracks the forward pass and associated fp8 metadata for easier debugging. I tried to repro the issue [here](https://github.com/esoba/recsys-examples/blob/debugging-with-repro/examples/hstu/minimal_repro_te.py) with some dummy inputs, and it has run successfully - have a hunch that there are NaN gradients flowing back into embedding table. 

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rapids-repo-template/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: examples/hstu/movielens_ranking_fp8.gin [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+RankingArgs.eval_metrics = ("AUC",)
+
+OptimizerArgs.optimizer_str = 'adam'
+OptimizerArgs.learning_rate = 1e-3
+OptimizerArgs.adam_beta1 = 0.9
+OptimizerArgs.adam_beta2 = 0.98
+
+TensorModelParallelArgs.tensor_model_parallel_size = 2
+
+# MixedPrecisionArgs.mixed_precision_dtype = "fp8"
```
Comment (esoba) [inline examples/hstu/movielens_ranking_fp8.gin:34]: Uncommenting this line should throw error at iteration 64 related to NaN loss

Comment (shijieliu) [reply examples/hstu/movielens_ranking_fp8.gin:34]: reproduce command
```
PYTHONPATH=${PYTHONPATH}:$(realpath ../) torchrun --nproc_per_node 2 --master_addr localhost --master_port 6000  pretrain_gr_ranking.py --gin-config-file movielens_ranking_fp8.gin
```

File: examples/hstu/modules/native_hstu_layer.py [hunk 2]
```diff
@@ -230,6 +234,12 @@ def forward(self, jd: JaggedData) -> JaggedData:
                 max_seqlen=jd.max_seqlen,
                 target_group_size=self._target_group_size,
             )
+            
+            # TODO: Remove this once the attention kernel outputs consistent dtype
```
Comment (shijieliu) [inline examples/hstu/modules/native_hstu_layer.py:238]: @JacoCheung could you double check why we need this?

Comment (JacoCheung) [reply examples/hstu/modules/native_hstu_layer.py:238]: When training with fp8 enabled, the model weight could be bf16/fp16 （NetworkArgs.dtype_str) (Usually it's bf16). So as activation is.
 
But the hstu kernel output is **fp16**,  so here we need a cast between fp16->bf16. @shijieliu 

@esoba do you think if there's a need to move the cast into the kernel or not?

Comment (esoba) [reply examples/hstu/modules/native_hstu_layer.py:238]: I think casting fp16 to bf16 wouldn't result in error since dynamic range is larger, but would assume some additional quantization error pops up (ideally that gets learned by model anyways). I think for consistency it would probably be better to move it into the kernel but as a workaround casting outside should be fine.

File: examples/hstu/modules/utils.py [hunk 3]
```diff
@@ -30,3 +37,38 @@ def init_mlp_weights_optional_bias(
         # Always initialize bias to zero.
         if m.bias is not None:
             m.bias.data.fill_(0.0)
+
+def convert_te_linear_to_torch_linear(m: torch.nn.Module) -> torch.nn.Module:
```
Comment (JacoCheung) [inline examples/hstu/modules/utils.py:41]: @esoba Is this used for debugging purpose or anything else?

Comment (esoba) [reply examples/hstu/modules/utils.py:41]: Sorry just debugging - added this when looking te compatible shapes/GEMMs

File: examples/hstu/modules/hstu_processor.py [hunk 4]
```diff
@@ -302,8 +304,111 @@ def forward(
...
 
+        # FP8 alignment: Truncate to nearest multiple of 16 for TE Linear fwd + bwdcompatibility
+        if not isinstance(self.config, InferenceHSTUConfig):
+            if self.config.fp8 is not None and self.training:
+                warnings.warn("Aligning JaggedData to nearest multiple of 16 for FP8 TE fwd + bwd compatibility")
+                jd = self._align_jagged_data_for_fp8(jd, fp8_alignment_mode=self.config.fp8_alignment_mode)
+
         return jd
 
+    def _align_jagged_data_for_fp8(
```
Comment (JacoCheung) [inline examples/hstu/modules/hstu_processor.py]: Hi @esoba , Since you have padding here, there should be a discarding process at the postprocessor  **before loss compute**. That's being said,
```
 final_loss = drop_pad_values(final_loss)
 final_loss.mean().backward()
```

Otherwise, the padded token will impact the backward both data gradient  and weight gradient **even if the padded value is initialized as 0.**

Comment (JacoCheung) [inline examples/hstu/modules/hstu_processor.py:315]: Hi @esoba , Since you have padding here, there should be a discarding process at the postprocessor  **before loss compute**. That's being said,
```
 final_loss = drop_pad_values(final_loss)
 final_loss.mean().backward()
```
Otherwise, the padded token will impact the backward both data gradient  and weight gradient **even if the padded value is initialized as 0.**


See [our loss calculation](https://github.com/esoba/recsys-examples/blob/main/examples/hstu/pipeline/train_pipeline.py#L843-L845). 

And our post-processor (if it's ranking) will take the padded token as normal token..

Comment (esoba) [reply examples/hstu/modules/hstu_processor.py:315]: I believe I was seeing the issue when I set this to truncate (cut off the last N elements to get nearest divisible by 16), let me double check this to see if there is any undefined behavior doing it this way as well. Thanks for the catch!

**Comment by JacoCheung** [comment]
FBGEMM_GPU HSTU has been integrated into recsys-example #321 .

FYI.
@XinboZhao  You can continue fp8 exploration based on current repo.
